### PR TITLE
Add sfinae function signature checks

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
This add several sfinae function signature checks. In order to accomplish without a lot of work it required upgrading to c++17.

Also includes a fix to tesseract collision example.

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
This PR should be ready for review. I forgot to add the function signature check to the command language.

File: README.md [hunk 1]
```diff
@@ -6,7 +6,6 @@ Platform             | CI Status
 ---------------------|:---------
 Linux (Focal)        | [![Build Status](https://github.com/ros-industrial-consortium/tesseract/workflows/Focal-Build/badge.svg)](https://github.com/ros-industrial-consortium/tesseract/actions)
 Linux (Bionic)       | [![Build Status](https://github.com/ros-industrial-consortium/tesseract/workflows/Bionic-Build/badge.svg)](https://github.com/ros-industrial-consortium/tesseract/actions)
-Linux (Xenial)       | [![Build Status](https://github.com/ros-industrial-consortium/tesseract/workflows/Xenial-Build/badge.svg)](https://github.com/ros-industrial-consortium/tesseract/actions)
```
Comment (mpowelson) [inline README.md:9]: Gasp!

**Review by mpowelson — COMMENTED** [review]
These are pretty confusing, but I think there are correct. Not something the average user will touch anyway.

**Comment by mpowelson** [comment]
We could also consider moving this into tesseract_planning if #421 ever happened. That might allow us to keep tesseract_core as a higher quality level than tesseract_planning since I think C++17 probably hurts us in that regard.

File: README.md (hunk 1 above)
Comment (Levi-Armstrong) [reply README.md:9]: > These are pretty confusing, but I think there are correct. Not something the average user will touch anyway.

They definitely are. It was a lot of trial and error.

**Comment by Levi-Armstrong** [comment]
@mpowelson Do you approve of the changes?

**Review by mpowelson — APPROVED** [review]

**Pull request merged**